php-checks/src/main/java/org/sonar/php/checks/UseOfUninitializedVariableCheck.java

```diff
@@ -87,175 +97,264 @@
+
+    Set<String> providedVariables = getParameterVariableNames(tree);
+    if (tree.is(Kind.FUNCTION_EXPRESSION)) {
+      providedVariables.addAll(getLexicalVariableNames((FunctionExpressionTree) tree));
+    }
 
-  private static class VariableVisitor extends PHPVisitorCheck {
+    InitialDataCollector initialDataCollector = new InitialDataCollector();
+    tree.accept(initialDataCollector);
+    // Catch clauses do not appear in the CFG. We collect all exception variables in the function body and
+    // consider them as provided in the whole function body to avoid false positives.
+    providedVariables.addAll(initialDataCollector.exceptionVariables);
 
-    final FunctionTree currentFunction;
+    blockSummaries.get(cfg.start()).predecessorsSummary.initializedVariables.addAll(providedVariables);
 
-    boolean trustVariables = true;
+    Deque<CfgBlock> workList = new ArrayDeque<>(cfgBlocks);
+    while (!workList.isEmpty()) {
+      CfgBlock block = workList.pop();
+      BlockSummary summary = blockSummaries.get(block);
 
-    Map<String, VariableIdentifierTree> firstVariableReadAccess = new HashMap<>();
+      for (CfgBlock successor : block.successors()) {
+        PredecessorsSummary predecessorsSummary = blockSummaries.get(successor).predecessorsSummary;
+        if (!predecessorsSummary.containsBlockSummary(summary)) {
+          predecessorsSummary.addBlockSummary(summary);
+          workList.add(successor);
+        }
+      }
+    }
+
+    Map<String, Set<Tree>> uninitializedVariableUses = new HashMap<>();
+    cfgBlocks.forEach(b -> checkBlock(b, blockSummaries.get(b))
+      .forEach((v, s) -> uninitializedVariableUses.computeIfAbsent(v, st -> new HashSet<>()).addAll(s)));
+
+    // static variables could be initialized after first usage. We only raise an issue on them when they were never
+    // initialized (i.e., not initialized by predecessors of the end block)
+    uninitializedVariableUses.entrySet().stream()
+      .filter(e -> !isInitializedStaticVariable(e.getKey(),
+        initialDataCollector.uninitializedStaticVariables,
+        blockSummaries.get(cfg.end())))
+      .forEach(e -> reportOnFirstTree(e.getValue()));
+  }
+
+  private static boolean isInitializedStaticVariable(String var, Set<String> uninitializedStaticVariables, BlockSummary endBlockSummary) {
+    return uninitializedStaticVariables.contains(var) && endBlockSummary.predecessorsSummary.wasInitialized(var);
+  }
 
-    Set<String> initializedVariables = new HashSet<>();
+  private void reportOnFirstTree(Set<Tree> trees) {
+    trees.stream()
+      .min(Comparator.comparingInt(a -> ((PHPTree) a).getFirstToken().line())
+        .thenComparing(a -> ((PHPTree) a).getFirstToken().column()))
+      .ifPresent(t -> newIssue(t, MESSAGE));
+  }
 
-    VariableVisitor(FunctionTree currentFunction) {
-      this.currentFunction = currentFunction;
+  private static boolean blockIsUnreachable(CfgBlock block, ControlFlowGraph cfg) {
```

Comment:
I think it would be easier and more efficient to have a method to compute the set of all reachable blocks and call it only once for a given CFG.